File: samples/FsBulletML.Sample/Particle.fs
namespace FsBulletML.Sample

open System
open Microsoft.Xna.Framework
open Microsoft.Xna.Framework.Graphics

type Particle (texture, pos, speed, angle, scale, shrinkRate, duration, color) as this =
  [<DefaultValue>]val mutable private texture : Texture2D 
  [<DefaultValue>]val mutable private pos : Vector2 
  [<DefaultValue>]val mutable private direction : Vector2 
  [<DefaultValue>]val mutable private originPos : Vector2 
  [<DefaultValue>]val mutable private duration : float32
  [<DefaultValue>]val mutable private scale : float32
  [<DefaultValue>]val mutable private shrinkRate : float32
  [<DefaultValue>]val mutable private speed : float32
  [<DefaultValue>]val mutable alive : bool
  [<DefaultValue>]val mutable private color : Color

  do
    this.texture <- texture
    this.pos <- pos
    this.scale <- scale
    this.shrinkRate <- shrinkRate
    this.alive <- true
    this.duration <- duration
    this.color <- color
    this.speed <- speed
    this.originPos <- new Vector2(texture.Width / 2 |> float32, texture.Height / 2 |> float32)
    let angle = MathHelper.ToRadians(angle)
    let up = new Vector2(0.f, -1.f)
    let rot = Matrix.CreateRotationZ(angle)
    this.direction <- Vector2.Transform(up, rot)

  member this.Update(delta:float) =
    let delta = float32 delta
    this.pos <- this.pos + this.direction * this.speed * delta
    this.scale <- this.scale - this.shrinkRate * delta
    this.duration <- this.duration - delta
     
    if (this.scale <= 0.f || this.duration <= 0.f) then
      this.alive <- false
      this.pos <- new Vector2(-100.f, -100.f)

  member this.Draw(sp:SpriteBatch) =
    sp.Draw(this.texture, this.pos, new Nullable<Rectangle>(), this.color, 0.f, this.originPos, this.scale, SpriteEffects.None, 0.f)

File: samples/FsBulletML.Sample.MonoGame.CSharp/Particle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace FsBulletML.Sample.MonoGame.CSharp
{
    class Particle
    {
        private Texture2D texture;
        private Vector2 position;
        private Vector2 direction;
        private Vector2 origin;
        private float duration;
        private float scale;
        private float shrinkRate;
        private float speed;
        private Color colour;
        public bool IsActive { get; set; }

        public Particle(Texture2D tex, Vector2 pos, float speed, float angle, float scale, float shrinkRate, float duration, Color colour)
        {
            this.texture = tex;
            this.position = pos;
            this.scale = scale;
            this.shrinkRate = shrinkRate;
            this.duration = duration;
            this.colour = colour;
            this.speed = speed;
            this.IsActive = true;

            this.origin = new Vector2(tex.Width / 2, tex.Height / 2);
            angle = MathHelper.ToRadians(angle);
            this.direction = Vector2.Transform(new Vector2(0, -1.0f), Matrix.CreateRotationZ(angle));
        }

        public void Update(double delta)
        {
            var d = (float)delta;
            this.position += this.direction * this.speed * d;
            this.scale -= this.shrinkRate * d;
            this.duration -= d;

            if (this.scale <= 0.0f || this.duration <= 0.0f)
            {
                this.IsActive = false;
                this.position = new Vector2(-100, -100);
            }
        }

        public void Draw(SpriteBatch sp)
        {
            sp.Draw(this.texture, this.position, null, this.colour, 0, this.origin, this.scale, SpriteEffects.None, 0);
        }
    }
}
